docs/src/developer/interfaces.md

```diff
@@ -0,0 +1,175 @@
+concrete multivariate type and for the generic multivariate polynomial
+type `MPoly{T}` only.
+
+## Generic interfaces
+
+As mentioned above, the generic implementations in Nemo depend on carefully
+written interfaces for each of the abstract types provided by the system.
+
+These interfaces are spelled out in the AbstractAlgebra documentation. Note
+that a generic implementation may depend on functions in both the required and
+optional interfaces as the optional functions are all implemented with generic
+fallbacks in terms of the required functions.
+
+For convenience we provide here a list of interfaces that can be relied on in
+generic implementations, along with a description.
+
+* Ring : all commutative rings in the system
+* Field : all fields in the system
+* NCRing : all rings in the system (not necessarily commutative)
+* Euclidean Ring : Euclidean rings (see notes below)
+* Univariate Polynomial Ring : all dense univariate polynomials
+* Multivariate Polynomial Ring : all sparse distributed multivariate polys.
+* Series Ring : all series, relative and absolute
+* Residue Ring : all quotients of gcd domains with `gcdx` by a principal ideal
+* Fraction Field : all fractions over a gcd domain with `gcdx`
+* Module : all finitely presented modules over a Euclidean domain
+* Matrix : all matrices over a commutative ring
+* Map : all (set) maps in the system
+
+Although we allow `Z/nZ` in our definition of Euclidean ring, much of the
+functionality in Nemo can be expected to misbehave (impossible inverses, etc.)
+when working with Euclidean rings that are not domains. In some cases the
+algorithms just don't exist, and in other cases we simply haven't implemented
+the required functionality to support all Euclidean rings for which
+computations can be done.
+
+Whether a ring is a Euclidean domain or not cannot be encoded in the type. Thus
+there is no abstract type for Euclidean domains or their elements. Instead,
+generic functions rely on the existence of certain functions such as `gcdx` to
+implement functionality for Euclidean domains.
+
+There is also currently no way to define a Euclidean function for a given ring
+(which is known to be Euclidean) and have the system recognise the ring as
+such. This kind of Euclidean interface may be provided in a future version of
+Nemo.
+
+
+## Julia interfaces we support
+
+Many Julia interfaces rely on being able to create zero and one elements given
+the type only. As we use the parent/element model (see developer notes on this
+topic) we cannot support all Julia interfaces fully.
+
+We do however partially implement some Julia interfaces.
+
+* Iteration : iterators are currently provided for multivariate polynomials to
+iterate over the coefficients, terms and monomials. Nemo matrices can also be
+iterated over. Iteration proceeds down each column in turn. One can also
+iterate over all permutations and partitions. Finally, the Flint `fq_nmod`
+finite field type can be iterated over.
```

Comment:
All finite fields can be iterated over (there is a generic implementation).

Reply:
Fixed.